communication/src/allocator/thread.rs

```diff
@@ -26,11 +26,14 @@ pub struct Thread {
 }
 
 impl Allocate for Thread {
+    type Pusher<T: Data> = ThreadPusher<Message<T>>;
+    type Puller<T: Data> = ThreadPuller<Message<T>>;
+
     fn index(&self) -> usize { 0 }
     fn peers(&self) -> usize { 1 }
-    fn allocate<T: 'static>(&mut self, identifier: usize) -> (Vec<Box<dyn Push<Message<T>>>>, Box<dyn Pull<Message<T>>>) {
+    fn allocate<T: Data>(&mut self, identifier: usize) -> (Vec<Self::Pusher<T>>, Self::Puller<T>) {
```

Comment:
Possibly a breaking change as it requires all datums to be `Data`, which is currently not required on the same thread.

Reply:
I don't know if this is the case because the trait itself defines this function as  `allocate<A: Data>`. I'm not sure why it was allowed to only specify `'static` here

https://docs.rs/timely_communication/0.12.0/timely_communication/allocator/trait.Allocate.html#tymethod.allocate

Reply:
Looked a bit more into this and it's definitely not a breaking change. The `: Data` bound was being previously silently enforced so better to spell it out explicitly.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4a5b3884be978fa9b05a4eb5aabc1be8